---
title: "PackToTheFuture"
output: rmarkdown::pdf_document
vignette: >
  %\VignetteIndexEntry{PackToTheFuture}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(PackToTheFuture)
```

### Introduction

The **`PackToTheFuture`** package is designed to streamline data-science workflows related to the Central Dogma of Molecular Biology applied in a bioinformatics contexts. Specifically, it provides five functions and one data object:

-   function 1 –\> `create_seq()`

-   function 2 –\> `get_rna()`

-   function 3 –\> `get_codons()`

-   function 4 –\> `traduction_prot()`

-   function 5 –\> `plot_base_seq()`

-   data object -\> `genetic_code`

In the following, we will go through each function in the package one by one and show how to use them together in a workflow. Finally we discuss typical use cases and potential future enhancements.

### Function 1 (`create_seq()`)

*Purpose*: Generate a random sequence composed of nucleotides A, T, G, and C.

Function definition:

``` r
create_seq <- function(seq_size){
  seq_list <- sample(c("A", "T", "G", "C"), size = seq_size, replace = TRUE)
  seq_str <- paste0(seq_list, collapse = "")
  return(seq_str)
}
```

#### Parameters

| Argument |  Type   |            Description             |
|:--------:|:-------:|:----------------------------------:|
| seq_size | numeric | Length of the sequence to generate |

#### Returns

A character string representing a DNA sequence.

#### Example usage

``` r
# Generate a sequence of length 12
seq1 <- create_seq(12)
seq1
```

Expected output format:

``` r
"ATGCGTACGTTA"
```

Each call produces a new random sequence of the same general format.

### Function 2

This function takes a DNA sequence and converts it into an RNA sequence by replacing "T" with "U".

#### Function definition

```{r}
get_rna <- function(dna_seq){
  rna_seq <- gsub("T", "U", dna_seq)
  return(rna_seq)
}
```

#### Parameters

dna_seq: the DNA sequence string

#### Returns

rna_seq: the RNA sequence string

#### Example usage

```{r}
get_rna("ACTG")
```

### Function 3: `get_codons()`

The `get_codons()` function takes a DNA sequence `seq` and splits it into triplet (codon) sections. The reading frame (starting position) is determined by the `start` argument. The generated codons can later be used to translate to amino acids.

```{r}
get_codons <- function(seq, start = 1){
  seq_len <- nchar(seq)
  codons <- substring(seq,
                      first = seq(from = start, to = seq_len-3+1, by = 3),
                      last = seq(from = 3+start-1, to = seq_len, by = 3))
  return(codons)
}
```

**Example usage:**

```{r}
DNAseq <- "ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG" 

# example 1: no reading frame is defined: start = 1 by default
codons_example1 <- get_codons(DNAseq)
#> [1] "ATG" "GCC" "ATT" "GTA" "ATG" "GGC" "CGC" "TGA" "AAG" "GGT" "GCC" "CGA" "TAG"

# example 2: shift the reading frame by one base: start = 2
codons_example2 <- get_codons(DNAseq, start = 2)
#> [1] "TGG" "CCA" "TTG" "TAA" "TGG" "GCC" "GCT" "GAA" "AGG" "GTG" "CCC" "GAT"
```

This function is particularly useful in bioinformatics workflows for:

-   Translating nucleotide sequences into amino acids (after codon extraction).

-   Exploring open reading frames (ORFs).

-   Checking for mutations that shift reading frames.

### Function 4 (`traduction_prot()`)

*Purpose*: Translate a vector of codons into a protein sequence using a predefined genetic code.

Function definition:

``` r
traduction_prot <- function(codons){
  prot_seq <- paste0(genetic_code[codons], collapse = "")
  return(prot_seq)
}
```

#### Parameters

| Argument |               Type               |           Description           |
|:-------------:|:-------------:|:----------------------------------------:|
|  codons  | character vector (of RNA codons) | The codon sequence to translate |

#### Returns

A concatenated protein string.

#### Example usage

``` r
cod_vec <- c("AUG", "ACG", "AAG")
traduction_prot(cod_vec)
```

#### Output:

``` r
"MTK"
```

### Function 5: plot_base_freq

A function that takes a sequence of nucleotides (DNA or RNA) and plots the number of occurences of each nucleotide in a barplot.

Packages are added in the DESCRIPTION file through `usethis::use_package()`. Also are added in the funcion's Roxygen Skeleton using importFrom and importing only the required function from the packages.

#### Parameters

seq: sequence of nucleotides as a string

#### Returns

ggplot barplot with the counts per nucleotide

#### Example usage

```{r}
plot_base_seq("AGCTAGCTAGCTCGGTTTCGTAGCTCATGC")
plot_base_seq("AUCGAUGGCUAGGCUGUAGCGAUACGUAGCGAUGCUA")
```
